File: reference/FlaMEVB/modBitmap.vb
Imports OpenTK.Graphics.OpenGL

Public Module modBitmap

    Public Function LoadBitmap(Path As String, ByRef ResultBitmap As Bitmap) As sResult
        Dim ReturnResult As sResult
        ReturnResult.Problem = ""
        ReturnResult.Success = False

        Dim Bitmap As Bitmap

        Try
            Bitmap = New Bitmap(Path)
        Catch ex As Exception
            ReturnResult.Problem = ex.Message
            ResultBitmap = Nothing
            Return ReturnResult
        End Try

        ResultBitmap = New Bitmap(Bitmap) 'copying the bitmap is needed so it doesn't lock access to the file

        ReturnResult.Success = True
        Return ReturnResult
    End Function

    Public Function SaveBitmap(Path As String, Overwrite As Boolean, BitmapToSave As Bitmap) As sResult
        Dim ReturnResult As sResult
        ReturnResult.Problem = ""
        ReturnResult.Success = False

        Try

            If IO.File.Exists(Path) Then
                If Overwrite Then
                    IO.File.Delete(Path)
                Else
                    ReturnResult.Problem = "File already exists."
                    Return ReturnResult
                End If
            End If
            BitmapToSave.Save(Path)

        Catch ex As Exception
            ReturnResult.Problem = ex.Message
            Return ReturnResult
        End Try

        ReturnResult.Success = True
        Return ReturnResult
    End Function

    Public Structure sBitmapGLTexture
        Public Texture As Bitmap
        Public TextureNum As Integer
        Public MipMapLevel As Integer
        Public MinFilter As TextureMinFilter
        Public MagFilter As TextureMagFilter

        Public Sub Perform()

            Dim BitmapData As Drawing.Imaging.BitmapData = Texture.LockBits(New Rectangle(0, 0, Texture.Width, Texture.Height), Imaging.ImageLockMode.ReadOnly, Imaging.PixelFormat.Format32bppArgb)

            If MipMapLevel = 0 Then
                GL.GenTextures(1, TextureNum)
            End If
            GL.BindTexture(TextureTarget.Texture2D, TextureNum)
            If MipMapLevel = 0 Then
                GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureWrapS, TextureWrapMode.ClampToEdge)
                GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureWrapT, TextureWrapMode.ClampToEdge)
                GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureMagFilter, MagFilter)
                GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureMinFilter, MinFilter)
            End If
            GL.TexImage2D(TextureTarget.Texture2D, MipMapLevel, PixelInternalFormat.Rgba8, Texture.Width, Texture.Height, 0, PixelFormat.Bgra, PixelType.UnsignedByte, BitmapData.Scan0)

            Texture.UnlockBits(BitmapData)
        End Sub
    End Structure

    Public Function BitmapIsGLCompatible(BitmapToCheck As Bitmap) As clsResult
        Dim ReturnResult As New clsResult("Compatability check")

        If Not SizeIsPowerOf2(BitmapToCheck.Width) Then
            ReturnResult.WarningAdd("Image width is not a power of 2.")
        End If
        If Not SizeIsPowerOf2(BitmapToCheck.Height) Then
            ReturnResult.WarningAdd("Image height is not a power of 2.")
        End If
        If BitmapToCheck.Width <> BitmapToCheck.Height Then
            ReturnResult.WarningAdd("Image is not square.")
        End If

        Return ReturnResult
    End Function
End Module

File: source/SharpFlame/modBitmap.cs
using System;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using OpenTK.Graphics.OpenGL;
using PixelFormat = System.Drawing.Imaging.PixelFormat;

namespace SharpFlame
{
    public sealed class modBitmap
    {
        public static modProgram.sResult LoadBitmap(string Path, ref Bitmap ResultBitmap)
        {
            modProgram.sResult ReturnResult = new modProgram.sResult();
            ReturnResult.Problem = "";
            ReturnResult.Success = false;

            Bitmap Bitmap = default(Bitmap);

            try
            {
                Bitmap = new Bitmap(Path);
            }
            catch ( Exception ex )
            {
                ReturnResult.Problem = ex.Message;
                ResultBitmap = null;
                return ReturnResult;
            }

            ResultBitmap = new Bitmap(Bitmap); //copying the bitmap is needed so it doesn't lock access to the file

            ReturnResult.Success = true;
            return ReturnResult;
        }

        public static modProgram.sResult SaveBitmap(string Path, bool Overwrite, Bitmap BitmapToSave)
        {
            modProgram.sResult ReturnResult = new modProgram.sResult();
            ReturnResult.Problem = "";
            ReturnResult.Success = false;

            try
            {
                if ( File.Exists(Path) )
                {
                    if ( Overwrite )
                    {
                        File.Delete(Path);
                    }
                    else
                    {
                        ReturnResult.Problem = "File already exists.";
                        return ReturnResult;
                    }
                }
                BitmapToSave.Save(Path);
            }
            catch ( Exception ex )
            {
                ReturnResult.Problem = ex.Message;
                return ReturnResult;
            }

            ReturnResult.Success = true;
            return ReturnResult;
        }

        public struct sBitmapGLTexture
        {
            public Bitmap Texture;
            public int TextureNum;
            public int MipMapLevel;
            public TextureMinFilter MinFilter;
            public TextureMagFilter MagFilter;

            public void Perform()
            {
                BitmapData BitmapData = Texture.LockBits(new Rectangle(0, 0, Texture.Width, Texture.Height),
                    ImageLockMode.ReadOnly, PixelFormat.Format32bppArgb);

                if ( MipMapLevel == 0 )
                {
                    GL.GenTextures(1, out TextureNum);
                }
                GL.BindTexture(TextureTarget.Texture2D, TextureNum);
                if ( MipMapLevel == 0 )
                {
                    GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureWrapS, (int)TextureWrapMode.ClampToEdge);
                    GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureWrapT, (int)TextureWrapMode.ClampToEdge);
                    GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureMagFilter, (int)MagFilter);
                    GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureMinFilter, (int)MinFilter);
                }
                GL.TexImage2D(TextureTarget.Texture2D, MipMapLevel, PixelInternalFormat.Rgba8, Texture.Width, Texture.Height, 0, OpenTK.Graphics.OpenGL.PixelFormat.Bgra,
                    PixelType.UnsignedByte, BitmapData.Scan0);

                Texture.UnlockBits(BitmapData);
            }
        }

        public static clsResult BitmapIsGLCompatible(Bitmap BitmapToCheck)
        {
            clsResult ReturnResult = new clsResult("Compatability check");

            if ( !modProgram.SizeIsPowerOf2(BitmapToCheck.Width) )
            {
                ReturnResult.WarningAdd("Image width is not a power of 2.");
            }
            if ( !modProgram.SizeIsPowerOf2(BitmapToCheck.Height) )
            {
                ReturnResult.WarningAdd("Image height is not a power of 2.");
            }
            if ( BitmapToCheck.Width != BitmapToCheck.Height )
            {
                ReturnResult.WarningAdd("Image is not square.");
            }

            return ReturnResult;
        }
    }
}